### PR TITLE
arch: build: interrupt: encode multi level IRQ with C macro magic

### LINKED ISSUE #62883: [build][interrupts] Configurable number of bits for multi-level interrupts
**Describe the bug**
#61050 introduced the support of configurable bitmasks for multi-level interrupt, however, Zephyr's [`gen_defines.py`](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/dts/gen_defines.py) script still expects each level to be 8, see: still hardcoding the number of bits for the levels, i.e.:

https://github.com/zephyrproject-rtos/zephyr/blob/1fc16e65656e742fbe0adc89e0fdf685c51525a5/scripts/dts/gen_defines.py#L462-L468

This results in incorrectly encoded IRQ number from DT macros like `DT_INST_IRQN`:

https://github.com/zephyrproject-rtos/zephyr/blob/1fc16e65656e742fbe0adc89e0fdf685c51525a5/drivers/serial/uart_ns16550.c#L1248

And may cause issues in build-time installed ISRs, i.e. IRQ collision / installing ISR to the wrong offset / mismatch between static and dynamic IRQN.

**To Reproduce**
1. Use anything after #61050 - cce530cae4cac4ac46d5c932267d8156377578bd
2. Use a board with `MULTI_LEVEL_INTERRUPTS` enabled
3. Configure the 1st level interrupt bits to anything other than 8

**Expected behavior**
Zephyr should build and work normally for any plausible value of 1st level interrupt bits

**Impact**
This can be a showstopper for boards that requires the 1st level interrupt bits to be something other than 8

**Environment (please complete the following information):**
 - Zephyr rev: cce530cae4cac4ac46d5c932267d8156377578bd

### PR DESCRIPTION
**Pull request by ycsin**
This PR encodes the Zephyr multi-level interrupt using C macro magic in the devicetree, rather than depending on the [`gen_defines.py`](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/dts/gen_defines.py) script during build:

https://github.com/zephyrproject-rtos/zephyr/blob/1fc16e65656e742fbe0adc89e0fdf685c51525a5/scripts/dts/gen_defines.py#L435-L441

https://github.com/zephyrproject-rtos/zephyr/blob/1fc16e65656e742fbe0adc89e0fdf685c51525a5/scripts/dts/gen_defines.py#L456-L469

https://github.com/zephyrproject-rtos/zephyr/blob/1fc16e65656e742fbe0adc89e0fdf685c51525a5/scripts/dts/gen_defines.py#L482-L485

This will resolve the [dt & kconfig cross dependency issue](https://github.com/zephyrproject-rtos/zephyr/pull/63131#issuecomment-1736886469) encountered in #63131, and fixes #62883

After this change, the `*_IRQ_IDX_x_VAL_irq` of a node with an interrupt parent is the local IRQ number of that node (raw value as specified in the devicetree, or ARM GIC IRQ encoded output from `gen_defines.py`), rather than being multi-level encoded. The encoding of the IRQN only happens in certain DT macros, and the right one has to be used for build-time connected ISR, i.e. with `IRQ_CONNECT`

Fixes #62883 

> [!IMPORTANT]
> Blocking #62907

> [!NOTE]  
> **Caveat**: The current approach only works for a maximum of 3 levels, I'm not sure if a while-loop equivalent like the implementation in `gen_defines.py` is possible with the macros, but it is what the Zephyr interrupt encoding scheme currently supports, see [Multi-level Interrupt handling](https://docs.zephyrproject.org/latest/kernel/services/interrupts.html#multi-level-interrupt-handling)

### CODE REVIEW

File: scripts/dts/gen_defines.py [hunk 1]
```diff
@@ -453,20 +453,20 @@ def map_arm_gic_irq_type(irq, irq_num):
...
-        # Look for interrupt controller parent until we have none
-        while irq_ctrl.interrupts:
-            irq_num = (irq_num + 1) << 8
-            if "irq" not in irq_ctrl.interrupts[0].data:
-                err(f"Expected binding for {irq_ctrl!r} to have 'irq' in "
-                    "interrupt-cells")
-            irq_num |= irq_ctrl.interrupts[0].data["irq"]
-            irq_ctrl = irq_ctrl.interrupts[0].controller
-        return irq_num
+    # def encode_zephyr_multi_level_irq(irq, irq_num):
```
Comment (cfriedt) [inline scripts/dts/gen_defines.py:456]: I would remove the commented-out lines. Architecturally speaking, I think this proposal is sound - it's better to do this in C via DT macros than to have part of it live in python.

File: scripts/dts/gen_defines.py [hunk 2]
```diff
@@ -482,7 +482,7 @@ def encode_zephyr_multi_level_irq(irq, irq_num):
             if cell_name == "irq":
                 if "arm,gic" in irq.controller.compats:
                     cell_value = map_arm_gic_irq_type(irq, cell_value)
-                cell_value = encode_zephyr_multi_level_irq(irq, cell_value)
+                # cell_value = encode_zephyr_multi_level_irq(irq, cell_value)
```
Comment (cfriedt) [inline scripts/dts/gen_defines.py:485]: Remove commented-out lines, please.

**Comment by cfriedt** [comment]
Also worth noting, is that this is actually critical to close residual bugs from using bitmaps to encode a variable number of interrupts per level in a multi-level interrupt context, since the magic number `8` was hard-coded all over the place (python, C, etc).

As is it, the code in `main` is broken without this.

**Comment by ycsin** [comment]
Force pushed a documented and cleaner version

**Comment by ycsin** [comment]
@galak @mbolivar-ampere @mbolivar could you please take a look?

File: include/zephyr/devicetree.h [hunk 3]
```diff
@@ -2409,6 +2409,70 @@
...
+ * See doc/kernel/services/interrupts.rst for details
+ */
+#define Z_DT_MULTI_LEVEL_IRQN(node_id, idx)                                                        \
+	COND_CODE_1(Z_DT_HAS_GPARENT_INTC(node_id), (Z_DT_IRQN_L3(node_id, idx)),                  \
+		    (COND_CODE_1(Z_DT_HAS_PARENT_INTC(node_id), (Z_DT_IRQN_L2(node_id, idx)),      \
+				 (Z_DT_IRQN_BY_IDX(node_id, idx)))))
+
+/**
+ * @brief Get the node's Zephyr interrupt number at index
+ * If `CONFIG_MULTI_LEVEL_INTERRUPTS` is enabled, the interrupt number at index will be
```
Comment (gmarull) [inline include/zephyr/devicetree.h:2466]:
```suggestion
 * If @kconfig{CONFIG_MULTI_LEVEL_INTERRUPTS} is enabled, the interrupt number at index will be
```

File: drivers/gpio/gpio_sifive.c [hunk 4]
```diff
@@ -376,7 +376,7 @@ DEVICE_DT_INST_DEFINE(0,
 		    &gpio_sifive_driver);
 
 #define		IRQ_INIT(n)					\
-IRQ_CONNECT(DT_INST_IRQ_BY_IDX(0, n, irq),			\
+IRQ_CONNECT(DT_IRQN_BY_IDX(DT_DRV_INST(0), n),			\
```
Comment (gmarull) [inline drivers/gpio/gpio_sifive.c:379]: usage of `DT_DRV_INST` means there's a missing `INST` version missing

**Review by gmarull — DISMISSED** [review]
haven't reviewed the implementation, just found some general issues. this is also missing tests.

File: include/zephyr/devicetree.h (hunk 3 above)
Comment (ycsin) [reply include/zephyr/devicetree.h:2466]: fixed

File: drivers/gpio/gpio_sifive.c (hunk 4 above)
Comment (ycsin) [reply drivers/gpio/gpio_sifive.c:379]: you're right, just added a `INST` version for this

**Review by cfriedt — DISMISSED** [review]

**Review by jgl-meta — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Comment by jhedberg** [comment]
Is the assignee correct for this one? @mbolivar-ampere could you please take a look?

**Review by mbolivar-ampere — DISMISSED** [review]
This is on my queue, just a few minor requests to start out with.

`Z_FOO` is a new naming convention in this header for internal names. I'd rather not add one more. Please stick to the `_INTERNAL` convention if you want to choose names that clearly indicate they're not supposed to be used outside of the header. The `Z_` prefix is the subject of argument elsewhere in naming land as far as I recall.

Second, it looks like this won't work past a certain level of nesting in the interrupt tree, which seems like a concern to me that is not addressed in the commit messages -- what's the story here?

**Comment by mbolivar-ampere** [comment]
@cfriedt sorry that I was unclear. `_INTERNAL` is a suffix.

```
% rg _INTERNAL include/zephyr/devicetree*
include/zephyr/devicetree.h
141:#define DT_PATH(...) DT_PATH_INTERNAL(__VA_ARGS__)
3018:	DT_NODE_HAS_STATUS_INTERNAL(node_id, status)
4238:#define DT_PATH_INTERNAL(...) \
4240:/** @brief DT_PATH_INTERNAL() helper: prepends _S_ to a node name
4286:#define DT_NODE_HAS_STATUS_INTERNAL(node_id, status) \
```

**Comment by ycsin** [comment]
> Second, it looks like this won't work past a certain level of nesting in the interrupt tree, which seems like a concern to me that is not addressed in the commit messages -- what's the story here?

While the `gen_defines.py` method supports the encoding of any number of interrupt levels, other parts of the Zephyr actually only support up to 3rd level:

https://github.com/zephyrproject-rtos/zephyr/blob/90b9809c6f5563c8a8f93310fa65fdb8f321bc06/include/zephyr/irq.h#L283-L307

https://github.com/zephyrproject-rtos/zephyr/blob/90b9809c6f5563c8a8f93310fa65fdb8f321bc06/scripts/build/gen_isr_tables.py#L18-L28

However I just found that the interrupts documentation mentioned :

> _providing support for up to **four** interrupt levels using this arch_

https://github.com/zephyrproject-rtos/zephyr/blob/90b9809c6f5563c8a8f93310fa65fdb8f321bc06/doc/kernel/services/interrupts.rst?plain=1#L78

As far as I can tell, there is only 3 levels, not sure if that's a typo? CC @rgundi

**Comment by cfriedt** [comment]
> As far as I can tell, there is only 3 levels, not sure if that's a typo? CC @rgundi 

That seems to be more of a generalization for example purposes. I think there are only 3 levels supported right now in Zephyr.

**Review by cfriedt — CHANGES_REQUESTED** [review]
If the C macro approach is preferable (and I think it is), can we also delete the python script in this PR?

**Comment by ycsin** [comment]
> If the C macro approach is preferable (and I think it is), can we also delete the python script in this PR?

IRQ encoding is just a small part of the `gen_defines.py` script, I think we can convert the `map_arm_gic_irq_type` encoding to C macros _later_ such that IRQ encoding is done solely via C macros and the script will just translate numbers in devicetree as-is to the generated C header

**Review by cfriedt — APPROVED** [review]

**Review by rgundi — APPROVED** [review]
LGTM

**Review by cfriedt — APPROVED** [review]

**Review by peter-mitsis — APPROVED** [review]

**Comment by cfriedt** [comment]
@mbolivar-ampere - would be good to get your stamp on this one.

**Comment by ycsin** [comment]
ping devicetree maintainers @mbolivar-ampere @galak

**Review by fkokosinski — APPROVED** [review]
LGTM from the perspective of https://github.com/zephyrproject-rtos/zephyr/pull/62907.

**Comment by cfriedt** [comment]
Sorry - I twitched and merged.

**Pull request merged**